File: sliding_window_pattern/max_average_subarray.py
from typing import List


class Solution:
    def findMaxAverage(self,nums:List[int], k:int)->float:
        if k > len(nums):
            return None
        max_average = 0
        window_sum = sum(nums[:k])
        max_average = window_sum / k
        for i in range(k, len(nums)):
            window_sum = window_sum - nums[i-k ] + nums[i]
            max_average = max(max_average, window_sum / k)
        return max_average

s = Solution()
print(s.findMaxAverage([1,12,-5,-6,50,3], 4)) # 12.75   
# Time complexity: O(n)
# Space complexity: O(1)

File: sliding_window_pattern/max_average_subarray.js
// leetcode 643. Maximum Average Subarray 
// Given an array consisting of n integers, find the contiguous subarray of given length k that has the maximum average value.
// And you need to output the maximum average value.
// Example 1:
// Input: [1,12,-5,-6,50,3], k = 4
// Output: 12.75
// Explanation: Maximum average is (12-5-6+50)/4 = 51/4 = 12.75
// Note:
// 1 <= k <= n <= 30,000.
// Elements of the given array will be in the range [-10,000, 10,000].
var findMaxAverage = function(nums, k) {
    if(nums.length <k) return null;
    let maxAverage = 0;
    let windowSum = 0;
    for(let i=0; i<k; i++){
        windowSum += nums[i]
    }
    maxAverage = windowSum/k;
    for(let i=k; i<nums.length; i++){
        windowSum = windowSum - nums[i-k] + nums[i]
        maxAverage = Math.max(maxAverage, windowSum/k)
    }
    return maxAverage;
};

console.log(findMaxAverage([1,12,-5,-6,50,3], 4)) // 12.75